torch/csrc/jit/codegen/cuda/evaluator_common.cpp

```diff
@@ -0,0 +1,481 @@
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      collectBufferSizes(into, ite->thenBody().exprs());
+      collectBufferSizes(into, ite->elseBody().exprs());
+    }
+  }
+}
+
+//! Kernel IR utility, collects all the kir symbolic
+//!  integers we will need at runtime, i.e. after the
+//!  generated cuda kernel has already been compiled.
+//!  The values are to be used for runtime logic, like
+//!  `computeLaunchparams`.
+std::vector<kir::Val*> collectRuntimeUsedIntegers(
+    Fusion* fusion,
+    GpuLower* lower) {
+  std::vector<kir::Val*> ret;
+
+  // Collect extent and integer inputs
+  for (auto val : fusion->usedMathVals()) {
+    auto kir_val = lower->lowerValue(val);
+    if (auto kir_tv = dynamic_cast<kir::TensorView*>(kir_val)) {
+      for (auto id : kir_tv->domain()->domain()) {
+        ret.push_back(id->extent());
+      }
+    } else if (val->isFusionInput()) {
+      if (kir_val->isA<kir::Int>()) {
+        ret.push_back(kir_val);
+      }
+    }
+  }
+
+  // Collect allocation sizes:
+  collectBufferSizes(ret, lower->kernel()->topLevelExprs());
+
+  return makeSortedEvaluationList(ret);
+}
+//! Fusion IR utility, collects all the fusionIR symbolic
+//!  integers we will need at runtime, i.e. after the
+//!  generated cuda kernel has already been compiled.
+//!  The values are to be used for runtime logic, like
+//!  `canSchedule` in heuristic look up.
+std::vector<Val*> collectRuntimeUsedIntegers(Fusion* fusion) {
+  std::vector<Val*> ret;
+
+  // Collect extent and integer inputs
+  for (auto val : fusion->usedMathVals()) {
+    if (auto tv = dynamic_cast<TensorView*>(val)) {
+      for (auto id : tv->domain()->domain()) {
+        ret.push_back(id->extent());
+      }
+    } else if (val->isFusionInput()) {
+      if (val->isA<Int>()) {
+        ret.push_back(val);
+      }
+    }
+  }
+
+  return makeSortedEvaluationList(ret);
+}
+
+} // namespace
+
+template <typename IRContext>
+void PrecomputedIntegersBase<IRContext>::intializeValueList(
+    typename IRContext::EVALUATOR_TYPE& const_evaluator,
+    const std::vector<IR_VAL*>& sorted_value_list) {
+  // Initialize workspace
+  num_of_values_ = sorted_value_list.size();
+  defined_ = std::vector<bool>(num_of_values_, false);
+  is_constant_ = std::vector<bool>(num_of_values_, false);
+  values_ = std::vector<int64_t>(num_of_values_, -1);
+
+  // Fill in constants and assign evaluator indices
+  for (int i = 0; i < num_of_values_; i++) {
+    // Use an expression evaluator to test if value is const
+    auto const_val = const_evaluator.evaluate(sorted_value_list[i]);
+    if (const_val.has_value()) {
+      is_constant_[i] = true;
+      values_[i] = const_val.value();
+    }
+    sorted_value_list[i]->setEvaluatorIndex(i);
+  }
+}
+
+template <typename IRContext>
+c10::optional<int64_t> PrecomputedIntegersBase<IRContext>::getMaybeValueFor(
+    const IR_VAL* val) {
+  auto index = val->evaluatorIndex();
```

Comment:
I assume having an `evaluatorIndex` buried in the Fusion IR means we can only have one evaluator per `Fusion` or `Kernel` is there a way to enforce that, instead of assume it?

Reply:
Yes this was enforced since each `evaluatorIndex` could only be set once so it would throw exceptions if a second evaluator is trying to bind the integers in the `Fusion`.